### PR TITLE
Eliminate Hybrid Machinery

### PR DESCRIPTION
**Pull request by varunagrawal**
Merge all the work done in `HybridFactorGraph` elimination:

- Split HybridFactorGraph in `NonlinearHybridFactorGraph` and `GaussianHybridFactorGraph`.
- Add `HybridEliminate` method to perform elimination on `GaussianHybridFactorGraph`s.
- Refactor `discrete` to have a better API and make it easy for using in `hybrid`.
- Changed the `DCFactor` paradigm so that `keys()` has both the continuous and the discrete keys.
- CI updates for removing warnings in Windows, and updating Python install on MacOS.

### CODE REVIEW

**Comment by dellaert** [comment]
Please merge develop into both feature/hybrid and this branch before I review.

File: gtsam/discrete/DecisionTree.h [hunk 1]
```diff
@@ -41,6 +41,7 @@ namespace gtsam {
    */
   template<typename L, typename Y>
   class DecisionTree {
+
```
Comment (dellaert) [inline gtsam/discrete/DecisionTree.h:44]: remove changes from all files in discrete (I count 3 files)

File: gtsam/hybrid/DCFactor.h [hunk 2]
```diff
@@ -50,11 +61,12 @@ class DCFactor : public Factor {
...
    */
   DCFactor(const KeyVector& continuousKeys, const DiscreteKeys& discreteKeys)
-      : Base(continuousKeys), discreteKeys_(discreteKeys) {}
+      : Base(AllKeys(continuousKeys, discreteKeys)),
+        discreteKeys_(discreteKeys) {}
 
   // NOTE unsure if needed?
   explicit DCFactor(const DiscreteKeys& discreteKeys)
-      : discreteKeys_(discreteKeys) {}
+      : DCFactor({}, discreteKeys) {}
```
Comment (dellaert) [inline gtsam/hybrid/DCFactor.h:69]: Let's remove

File: gtsam/hybrid/DCFactor.h [hunk 3]
```diff
@@ -99,10 +111,24 @@ class DCFactor : public Factor {
    * @param discreteVals - Likewise, assignment to the discrete variables in
    * `discreteKeys__`.
    */
-  virtual boost::shared_ptr<GaussianFactor> linearize(
+  virtual GaussianFactor::shared_ptr linearize(
```
Comment (dellaert) [inline gtsam/hybrid/DCFactor.h:114]: revert and forward declare GaussianFactor, so we don't have to include GaussianFactor.h and co here.

File: gtsam/hybrid/DCFactor.h [hunk 4]
```diff
@@ -99,10 +111,24 @@ class DCFactor : public Factor {
...
       const Values& continuousVals,
       const DiscreteValues& discreteVals) const = 0;
 
+  /**
+   * @brief Linearize all the continuous factors only with respect to the
+   * continuous variables (as given in `keys_`).
+   *
+   * This `linearize` is different from the `linearize(continuous, discrete)` in
+   * that it assumes no assignment of discrete keys and linearizes all
```
Comment (dellaert) [inline gtsam/hybrid/DCFactor.h:123]: Comment on the return type. In fact, should it also be a GaussianMixture?

File: gtsam/hybrid/DCMixtureFactor.h [hunk 5]
```diff
@@ -33,55 +33,92 @@ namespace gtsam {
...
+                  const std::vector<sharedFactor>& factors,
+                  bool normalized = false)
+      : DCMixtureFactor(keys, discreteKeys,
+                        Factors(discreteKeys, factors)) {}
 
-  DCMixtureFactor& operator=(const DCMixtureFactor& rhs) {
-    Base::operator=(rhs);
-    this->dk_ = rhs.dk_;
-    this->factors_ = rhs.factors_;
+  /**
```
Comment (dellaert) [inline gtsam/hybrid/DCMixtureFactor.h:86]: Let's just kill. Caller should create shared pointers without copying

File: gtsam/hybrid/DCMixtureFactor.h [hunk 6]
```diff
@@ -33,55 +33,92 @@ namespace gtsam {
...
+      auto f_ptr = boost::make_shared<NonlinearFactorType>(factor);
+      factor_pointers.push_back(f_ptr);
+    }
+    // Generate decision tree based on discreteKey to factor mapping.
+    factors_ = Factors(discreteKeys, factor_pointers);
   }
 
   ~DCMixtureFactor() = default;
 
   double error(const Values& continuousVals,
```
Comment (dellaert) [inline gtsam/hybrid/DCMixtureFactor.h:105]: doxygen

File: gtsam/hybrid/GaussianMixture.h [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+   * @param discreteKeys - the keys for *discrete* variables
+   * @param conditionals A decision tree of GaussianConditional instances.
+   * TODO(Frank): (possibly wrongly) assumes nrFrontals is one
+   * TODO(Frank): should pass frontal keys, cont. parent keys, discrete parent
+   * keys, cannot be a conditional on a discrete key.
+   * TODO: desired API =
+   * GaussianConditionalMixture(const Conditionals& conditionals, 
+   *                            const DiscreteKeys& discreteParentKeys)
+   */
+  GaussianMixture(size_t nrFrontals,
```
Comment (dellaert) [inline gtsam/hybrid/GaussianMixture.h:67]: move definition to cpp? Too ugly here

File: gtsam/hybrid/GaussianMixture.h [hunk 8]
```diff
@@ -0,0 +1,116 @@
...
+                          % nrFrontals % p->nrFrontals()).str());
+                return boost::dynamic_pointer_cast<GaussianFactor>(p);
+              })),
+        BaseConditional(nrFrontals) {}
+
+  /// @}
+  /// @name Standard API
+  /// @{
+
+  GaussianConditional::shared_ptr operator()(
```
Comment (dellaert) [inline gtsam/hybrid/GaussianMixture.h:89]: same -> cpp

File: gtsam/slam/BetweenFactor.h [hunk 9]
```diff
@@ -80,10 +80,11 @@ namespace gtsam {
     /// @{
 
     /// print with optional string
-    void print(const std::string& s, const KeyFormatter& keyFormatter = DefaultKeyFormatter) const override {
-      std::cout << s << "BetweenFactor("
-          << keyFormatter(this->key1()) << ","
-          << keyFormatter(this->key2()) << ")\n";
+    void print(
```
Comment (dellaert) [inline gtsam/slam/BetweenFactor.h:83]: Remove from this PR. Core method of core factor should not be slipped in.

File: python/gtsam/tests/test_BetweenFactor.py [hunk 10]
```diff
@@ -0,0 +1,33 @@
+"""
```
Comment (dellaert) [inline python/gtsam/tests/test_BetweenFactor.py:1]: I don't see change to a .i file. ?

**Review by dellaert — COMMENTED** [review]
One more round and we can merge. Not yet approving as I feel strongly about some things

File: python/gtsam/tests/test_BetweenFactor.py (hunk 10 above)
Comment (varunagrawal) [reply python/gtsam/tests/test_BetweenFactor.py:1]: That change is already in develop for some reason.

File: gtsam/hybrid/DCFactor.h (hunk 4 above)
Comment (varunagrawal) [reply gtsam/hybrid/DCFactor.h:123]: DCFactor is the right type since it is the polymorphic base type.

File: gtsam/hybrid/DCFactor.h (hunk 3 above)
Comment (varunagrawal) [reply gtsam/hybrid/DCFactor.h:114]: I don't quite understand this. GaussianFactor.h is not included here, so do you want to remove the #include from the DCGaussianMixtureFactor?

File: gtsam/hybrid/DCMixtureFactor.h (hunk 5 above)
Comment (varunagrawal) [reply gtsam/hybrid/DCMixtureFactor.h:86]: I also found a bug in the `vector<sharedFactor>` version of the constructor. `normalized_` was not being set!

File: gtsam/hybrid/DCFactor.h (hunk 3 above)
Comment (dellaert) [reply gtsam/hybrid/DCFactor.h:110]: If it is not included, it should not compile. If it is included, it should be removed from this file.

File: gtsam/hybrid/DCFactor.h (hunk 4 above)
Comment (dellaert) [reply gtsam/hybrid/DCFactor.h:119]: I don't understand this, can it be anything else, ever? Make the type as specific as it can be.

**Comment by dellaert** [comment]
Don't worry about windows now. Will deal with windows when we try to merge feature/hybrid

**Comment by varunagrawal** [comment]
I actually did a normal push but GitHub didn't reflect it (some sort of bug on their end?) so I had to force push to get it to actually reflect.

File: gtsam/hybrid/DCFactor.h (hunk 3 above)
Comment (varunagrawal) [reply gtsam/hybrid/DCFactor.h:110]: Okay then there's nothing to do here since we include `GaussianFactorGraph.h` (which is needed) and that adds `GaussianFactor` as well. :)

File: gtsam/hybrid/DCFactor.h (hunk 4 above)
Comment (varunagrawal) [reply gtsam/hybrid/DCFactor.h:119]: So we can use DCGaussianMixtureFactor as the return type, but we've been using DCFactor since it is the common base class for all the DC factors. If we update this to use DCGaussianMixtureFactor, then it becomes a little messy since the forward declaration doesn't exactly work and there will be circular includes if we import DCGaussianMixtureFactor in DCFactor.

Comment (dellaert) [reply gtsam/hybrid/DCFactor.h:119]: Did you actually investigate or are you shooting from the hip? I think it's a valid review comment.
Why can you not just say 
```
class DCGaussianMixtureFactor; // forward declaration
```
above?  It should return a shared pointer, so really `boost::shared_ptr<DCGaussianMixtureFactor>`  should just work.

File: gtsam/hybrid/DCFactor.h (hunk 3 above)
Comment (dellaert) [reply gtsam/hybrid/DCFactor.h:110]: ok

**Review by dellaert — APPROVED** [review]
I'll approve so we can merge but that one comment is still valid I think. Code for the now, not the future. If it works now, do it.

File: gtsam/hybrid/DCFactor.h (hunk 4 above)
Comment (varunagrawal) [reply gtsam/hybrid/DCFactor.h:119]: I actually ran it with the forward declaration and got a bunch of compiler erros, but I did a `DCGaussianMixtureFactor::shared_ptr` which is the wrong thing to do. Let me try again and resolve it with full certainty.

Comment (varunagrawal) [reply gtsam/hybrid/DCFactor.h:119]: That worked!! Yay
